teps/0131-tekton-conformance-policy.md

```diff
@@ -0,0 +1,251 @@
+
+#### 2 Tekton Conformance Policy (this doc) v.s. [Tekton API Compatibility Policy][Tekton API Compatibility Policy].
+
+The two policies resemble each other in many aspects: they both define Tekton API fields, breaking changes, and  the practice of making API changes. However, in essence, they are different:
+
+The ***Conformance Policy*** defines ***Tekton Conformance API Specification*** that **vendors** MUST implement.
+* It only includes **Tekton GA Primitives**.
+* **Tekton Open Source Community** is like a "vendor" that provides the community version of Tekton "product" that contains a reference implementation that conforms to the Conformance Policy.
+
+The **Compatibility Policy** defines rules to follow when making changes on **Tekton APIs**, so that Tekton users can use it to define their CI/CD Pipeline.
+
+On the other hand, **Tekton APIs** drive the **Tekton Conformance APIs**, i.e. fields are added to the **Tekton Conformance APIs** as they graduate from **Tekton APIs**. The lifecycle of a field is:
+* Added to Tekton APIs as alpha
+* Updated to beta
+* Updated to GA (become an OPTIONAL field to Conformance APIs)
+* Added to Conformance APIs as REQUIRED
+
+Once the Conformance Policy is defined, the Tekton Compatibility Policy needs to be updated accordingly:
+
+Making backward compatible updates (in the context of Tekton Compatibility Policy)  on a Tekton API fields that's in the Conformance API (for example, adding a new field in an existing object field) is a backward incompatible change for Tekton Conformance Policy, which requires a new Conformance Policy Version (which needs to be approved by 50%+ community owners)
+
+#### 3. Do we only include GA primitives in the conformance policy?
+Yes, because GA primitives are reliable – making breaking changes for Tekton GA APIs are exceedingly costly. Vendors wouldn't want to invest a lot on APIs that are not stable enough.
+
+#### 4. Should we define the Conformance Policy per Tekton Services?
+I.e. Tekton Pipelines Conformance Policy, Tekton Triggers Conformance Policy, Tekton Results Conformance Policy.
+
+Yes because those services can exist independently, a vendor doesn't have to support all services to claim the Conformance.
+
+### Open Questions
+
+#### 1. How long do we want the policy update notice to be ahead of?
+Promoting a field to required is considered a breaking change for the Conformance Policy: vendors' conformant services become nonconformant on the new version of the conformance where a new field is promoted to "required". Therefore, we need to notify vendors ahead of time.
+
+9 months, open to suggestions.
+
+#### 2. Do we consider the previous Conformance version valid after the new version is released?
+Yes, the previous version should still be valid.
+
+#### 3. What to do with field deprecation?
```

Comment:
Can you expand a bit on this? I think if we deprecated a field that's optional for conformance, we wouldn't need to update the conformance version, but if we deprecated a required field we would.

Reply:
Updated! wdyt?

Reply:
Maybe you could provide an example timeline, like:
We want to deprecate a field that's optional for users but required for conformance
- We first make it optional for conformance. Does that require a policy bump? I don't understand why it's a breaking change.
- Next we deprecate it; it's still optional for conformance. I would think that doesn't require a policy bump. Vendors can remove support.
- field is removed from the api and conformance spec. Does this require a policy bump?

Reply:
Hi Lee,  thank you for the comment!

Making a required field in Conformance API as optional for conformance means vendors have the freedom to not support it. Say I have a previously conformant yaml:

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
name: "conformant"
spec:
  params:
  - name: "greet"
    value: "hello"
```

Then the vendor stops supporting the `params` field. That's said, the `param` field is unknown to vendors, so this conformant yaml becomes invalid, because the vendor service does not understand what `param` is.

Reply:
Thanks, that's helpful! It sounds like the policy bump is required at the time when a field that was required in the conformance spec becomes optional for conformance?

Reply:
Thanks, that's helpful! It sounds like the policy bump is required at the time when a field that was required in the conformance spec becomes optional for conformance?

Reply:
Yes, that's true, updated to doc to make it clearer!